samples/core/multi-device/main.cpp

```diff
@@ -0,0 +1,501 @@
+                               "y", "y_dim", "y dimension of input", false,
+                               4'096, "positive integral"));
+}
+template <>
+ConvolutionOptions cl::sdk::comprehend<ConvolutionOptions>(
+    std::shared_ptr<TCLAP::ValueArg<cl_uint>> x_dim_arg,
+    std::shared_ptr<TCLAP::ValueArg<cl_uint>> y_dim_arg)
+{
+    return ConvolutionOptions{ x_dim_arg->getValue(), y_dim_arg->getValue() };
+}
+
+// Host-side implementation of the convolution for verification. Padded input
+// assumed.
+void host_convolution(const std::vector<cl_float> in,
+                      std::vector<cl_float>& out,
+                      const std::vector<cl_float> mask, const cl_uint x_dim,
+                      const cl_uint y_dim)
+{
+    constexpr cl_uint mask_dim = 3;
+    constexpr cl_uint pad_width = mask_dim / 2;
+    const cl_uint in_dim_x = x_dim + pad_width * 2;
+
+    for (cl_uint gid_x = 0; gid_x < x_dim; ++gid_x)
+    {
+        for (cl_uint gid_y = 0; gid_y < y_dim; ++gid_y)
+        {
+            float result = 0.f;
+            for (cl_uint y = 0; y < mask_dim; ++y)
+            {
+                for (cl_uint x = 0; x < mask_dim; ++x)
+                {
+                    result += mask[y * mask_dim + x]
+                        * in[(gid_y + y) * in_dim_x + (gid_x + x)];
+                }
+            }
+            out[gid_y * x_dim + gid_x] = result;
+        }
+    }
+}
+
+int main(int argc, char* argv[])
+{
+    try
+    {
+        // Parse command-line options.
+        auto opts = cl::sdk::parse_cli<cl::sdk::options::Diagnostic,
+                                       cl::sdk::options::SingleDevice,
+                                       ConvolutionOptions>(argc, argv);
+        const auto& diag_opts = std::get<0>(opts);
+        const auto& dev_opts = std::get<1>(opts);
+        const auto& conv_opts = std::get<2>(opts);
+
+        // Create runtime objects based on user preference or default.
+        cl::Device dev = cl::sdk::get_context(dev_opts.triplet)
+                             .getInfo<CL_CONTEXT_DEVICES>()
+                             .at(0);
+
+        cl::Platform platform{
+            dev.getInfo<CL_DEVICE_PLATFORM>()
+        }; // https://github.com/KhronosGroup/OpenCL-CLHPP/issues/150
+
+        if (!diag_opts.quiet)
+        {
+            std::cout << "Selected device: " << dev.getInfo<CL_DEVICE_NAME>()
+                      << "\n"
+                      << "from " << platform.getInfo<CL_PLATFORM_VENDOR>()
+                      << " platform\n"
+                      << std::endl;
+        }
+
+        if (diag_opts.verbose)
+        {
+            std::cout << "Creating sub-devices...";
+            std::cout.flush();
+        }
+
+#if CL_HPP_TARGET_OPENCL_VERSION < 120
+        std::cerr
+            << "Error: OpenCL subdevices not supported before version 1.2 "
+            << std::endl;
+        exit(EXIT_FAILURE);
+#endif
```

Comment:
See comment above - I don't think we need this check.

Reply:
I still don't think we need this check.  If you want to try to be sure though, remove this code, and try to build for CL_HPP_TARGET_OPENCL_VERSION == 110 or lower, and it shouldn't build because it won't find the APIs for creating sub-devices.

Perhaps this is intended to be a runtime check based on the device version instead?